### PR TITLE
Fix: Research table in LL parks imported incompletely

### PR DESCRIPTION
**Pull request by Gymnasiast**
The code now imports up to 250 items (but will stop reading when it encounters the end-of-list-marker, like RCT1 does). The highest I have seen is 217, but I wanted to make some allowance for track part inventions and the like.

### CODE REVIEW

**Review by tupaschoal — COMMENTED** [review]
I don't think I understand how t his fixes it, would you care to elaborate a bit, please?

**Comment by Gymnasiast** [comment]
Sure. We assumed that the research list in RCT1 saves/scenarios had a maximum of 180 entries. Investigation of scenarios like Blackpool Pleasure Beach has revealed that this is not correct, as I have found saves with up to 217 research entries.

The reason this went undetected is because 1) most scenarios have 180 entries or fewer, and 2) the research items are followed by a large block of stuff we don’t import, so you would never know unless you were investigating.

Testing-wise, if you compare my version against develop, you will notice that in the fixed version, the Urban Theming is correctly put on the research list, while it isn’t on `develop`.

For reference, here is a printout of RCT-PHP, which how I found this problem in the first place:

```
Invented items:
Scenery entry: item GENERAL
Vehicle entry: ride type STEEL_CORKSCREW_ROLLER_COASTER, vehicle type CORKSCREW_ROLLER_COASTER_TRAIN
Vehicle entry: ride type STEEL_TWISTER_ROLLER_COASTER, vehicle type STEEL_TWISTER_ROLLER_COASTER_TRAIN
Vehicle entry: ride type BOAT_HIRE, vehicle type ROWING_BOATS
Vehicle entry: ride type BOAT_HIRE, vehicle type CANOES
Vehicle entry: ride type CAR_RIDE, vehicle type RACING_CARS
Vehicle entry: ride type CAR_RIDE, vehicle type SPORTSCARS
Vehicle entry: ride type CAR_RIDE, vehicle type TRUCKS
Vehicle entry: ride type CAR_RIDE, vehicle type CAT_CARS
Vehicle entry: ride type CAR_RIDE, vehicle type VINTAGE_CARS
Vehicle entry: ride type CHAIRLIFT, vehicle type CHAIRLIFT_CARS
Vehicle entry: ride type CHAIRLIFT, vehicle type CHAIRLIFT_CARS_ALTERNATIVE
Vehicle entry: ride type _3D_CINEMA, vehicle type CINEMA_BUILDING
Vehicle entry: ride type CIRCUS, vehicle type CIRCUS_TENT
Vehicle entry: ride type CROOKED_HOUSE, vehicle type CROOKED_HOUSE_BUILDING
Vehicle entry: ride type CYCLE_RAILWAY, vehicle type BICYCLES
Vehicle entry: ride type DODGEMS, vehicle type DODGEMS
Vehicle entry: ride type FERRIS_WHEEL, vehicle type FERRIS_WHEEL
Vehicle entry: ride type FLYING_ROLLER_COASTER, vehicle type FLYING_ROLLER_COASTER_TRAIN
Vehicle entry: ride type FLYING_SAUCERS, vehicle type FLYING_SAUCERS
Vehicle entry: ride type GO_KARTS, vehicle type GO_KARTS
Vehicle entry: ride type GHOST_TRAIN, vehicle type GHOST_TRAIN_CARS
Vehicle entry: ride type ROTO_DROP, vehicle type ROTODROP_CAR
Vehicle entry: ride type HAUNTED_HOUSE, vehicle type HAUNTED_HOUSE_BUILDING
Vehicle entry: ride type LOG_FLUME, vehicle type LOG_FLUME_BOATS
Vehicle entry: ride type MERRY_GO_ROUND, vehicle type MERRY_GO_ROUND
Vehicle entry: ride type MINE_TRAIN_ROLLER_COASTER, vehicle type MINE_TRAIN
Vehicle entry: ride type MINIATURE_GOLF, vehicle type GOLFERS
Vehicle entry: ride type MINIATURE_RAILWAY, vehicle type STEAM_TRAIN
Vehicle entry: ride type MINIATURE_RAILWAY, vehicle type STEAM_TRAIN_COVERED_CARS
Vehicle entry: ride type MONORAIL, vehicle type SMALL_MONORAIL_CARS
Vehicle entry: ride type MONORAIL, vehicle type LARGE_MONORAIL_TRAIN
Vehicle entry: ride type OBSERVATION_TOWER, vehicle type ROTATING_CABIN
Vehicle entry: ride type CYCLE_MONORAIL, vehicle type HELICOPTER_CARS
Vehicle entry: ride type RIVER_RAPIDS, vehicle type RAPIDS_BOATS
Vehicle entry: ride type REVERSE_FREEFALL_ROLLER_COASTER, vehicle type REVERSE_FREEFALL_ROLLER_COASTER_CAR
Vehicle entry: ride type WOODEN_REVERSER_ROLLER_COASTER, vehicle type REVERSER_CARS
Vehicle entry: ride type MOTION_SIMULATOR, vehicle type SIMULATOR_POD
Vehicle entry: ride type RIVER_RIDE, vehicle type RIVER_RIDE_BOATS
Vehicle entry: ride type STEEL_ROLLER_COASTER, vehicle type STEEL_ROLLER_COASTER_TRAIN
Vehicle entry: ride type STEEL_ROLLER_COASTER, vehicle type STEEL_ROLLER_COASTER_TRAIN_BACKWARDS
Vehicle entry: ride type WOODEN_SIDE_FRICTION_ROLLER_COASTER, vehicle type WOODEN_SIDE_FRICTION_CARS
Vehicle entry: ride type SPACE_RINGS, vehicle type SPACE_RINGS
Vehicle entry: ride type LAUNCHED_FREEFALL, vehicle type LAUNCHED_FREEFALL_CAR
Vehicle entry: ride type STEEL_MINI_ROLLER_COASTER, vehicle type LADYBIRD_CARS
Vehicle entry: ride type STEEL_MINI_ROLLER_COASTER, vehicle type LOG_CARS
Vehicle entry: ride type STEEL_MINI_ROLLER_COASTER, vehicle type ROCKET_CARS
Vehicle entry: ride type STEEL_MINI_ROLLER_COASTER, vehicle type SPINNING_CARS
Vehicle entry: ride type INVERTED_ROLLER_COASTER, vehicle type INVERTED_COASTER_TRAIN
Vehicle entry: ride type SUSPENDED_ROLLER_COASTER, vehicle type SUSPENDED_SWINGING_CARS
Vehicle entry: ride type SUSPENDED_ROLLER_COASTER, vehicle type SUSPENDED_SWINGING_AIRPLANE_CARS
Vehicle entry: ride type STEEL_WILD_MOUSE_ROLLER_COASTER, vehicle type STEEL_MOUSE_CARS
Vehicle entry: ride type SINGLE_RAIL_ROLLER_COASTER, vehicle type HORSES
Vehicle entry: ride type SINGLE_RAIL_ROLLER_COASTER, vehicle type MOTORBIKES
Vehicle entry: ride type SUSPENDED_MONORAIL, vehicle type SUSPENDED_MONORAIL_TRAIN
Vehicle entry: ride type SUSPENDED_SINGLE_RAIL_ROLLER_COASTER, vehicle type SINGLE_PERSON_SWINGING_CHAIRS
Vehicle entry: ride type SUSPENDED_SINGLE_RAIL_ROLLER_COASTER, vehicle type LYING_DOWN_SWINGING_CARS
Vehicle entry: ride type SWINGING_SHIP, vehicle type SWINGING_SHIP
Vehicle entry: ride type SWINGING_INVERTER_SHIP, vehicle type SWINGING_INVERTER_SHIP
Vehicle entry: ride type STAND_UP_STEEL_ROLLER_COASTER, vehicle type STANDUP_ROLLER_COASTER_CARS
Vehicle entry: ride type TOP_SPIN, vehicle type TOPSPIN_CAR
Vehicle entry: ride type BOBSLED_ROLLER_COASTER, vehicle type BOBSLEIGH_CARS
Vehicle entry: ride type TWIST, vehicle type TWIST_ARMS_AND_CARS
Vehicle entry: ride type HEARTLINE_TWISTER_ROLLER_COASTER, vehicle type HEARTLINE_TWISTER_CARS
Vehicle entry: ride type HEARTLINE_TWISTER_ROLLER_COASTER, vehicle type HEARTLINE_TWISTER_CARS_REVERSED
Vehicle entry: ride type VERTICAL_ROLLER_COASTER, vehicle type VERTICAL_ROLLER_COASTER_CARS
Vehicle entry: ride type VIRGINIA_REEL, vehicle type VIRGINIA_REEL_TUBS
Vehicle entry: ride type WATER_SLIDE, vehicle type DINGHIES
Vehicle entry: ride type WOODEN_WILD_MOUSE_ROLLER_COASTER, vehicle type WOODEN_MOUSE_CARS
Vehicle entry: ride type WOODEN_WILD_MOUSE_ROLLER_COASTER, vehicle type WOODEN_MINE_CARS
Vehicle entry: ride type WOODEN_ROLLER_COASTER, vehicle type WOODEN_ROLLER_COASTER_TRAIN
Vehicle entry: ride type WOODEN_ROLLER_COASTER, vehicle type WOODEN_ROLLER_COASTER_TRAIN_BACKWARDS
Vehicle entry: ride type WOODEN_TWISTER_ROLLER_COASTER, vehicle type WOODEN_TWISTER_ROLLER_COASTER_TRAIN
Ride improvement: type VERTICAL_LOOP, ride type STAND_UP_STEEL_ROLLER_COASTER
Ride improvement: type CORKSCREW, ride type STAND_UP_STEEL_ROLLER_COASTER
Ride improvement: type HALF_LOOP, ride type STAND_UP_STEEL_ROLLER_COASTER
Ride improvement: type VERTICAL_LOOP, ride type INVERTED_ROLLER_COASTER
Ride improvement: type INLINE_TWIST, ride type INVERTED_ROLLER_COASTER
Ride improvement: type CORKSCREW, ride type INVERTED_ROLLER_COASTER
Ride improvement: type HALF_LOOP, ride type INVERTED_ROLLER_COASTER
Ride improvement: type VERTICAL_LOOP, ride type STEEL_ROLLER_COASTER
Ride improvement: type CORKSCREW, ride type STEEL_CORKSCREW_ROLLER_COASTER
Ride improvement: type VERTICAL_LOOP, ride type STEEL_CORKSCREW_ROLLER_COASTER
Ride improvement: type HALF_LOOP, ride type STEEL_CORKSCREW_ROLLER_COASTER
Vehicle entry: ride type STEEL_TWISTER_ROLLER_COASTER, vehicle type STAND_UP_STEEL_TWISTER_ROLLER_COASTER_TRAIN
Vehicle entry: ride type STEEL_TWISTER_ROLLER_COASTER, vehicle type FLOORLESS_STEEL_TWISTER_ROLLER_COASTER_TRAIN
Vehicle entry: ride type STEEL_TWISTER_ROLLER_COASTER, vehicle type NON_LOOPING_STEEL_TWISTER_ROLLER_COASTER_TRAIN
Vehicle entry: ride type BOAT_HIRE, vehicle type SWANS_PEDAL_BOATS
Vehicle entry: ride type BOAT_HIRE, vehicle type WATER_TRICYCLES
Ride improvement: type VERTICAL_LOOP, ride type VERTICAL_ROLLER_COASTER
Vehicle entry: ride type BOAT_HIRE, vehicle type BUMPER_BOATS
Ride improvement: type WATER_SPLASH, ride type WOODEN_ROLLER_COASTER
Ride improvement: type LAUNCHED_LIFT_HILL, ride type STEEL_TWISTER_ROLLER_COASTER
Ride improvement: type LARGE_HALF_LOOP, ride type STEEL_TWISTER_ROLLER_COASTER
Ride improvement: type REVERSER_TURNTABLE, ride type LOG_FLUME
Ride improvement: type VERTICAL_LOOP, ride type WOODEN_ROLLER_COASTER
Vehicle entry: ride type STEEL_CORKSCREW_ROLLER_COASTER, vehicle type HYPERCOASTER_TRAIN
Vehicle entry: ride type INVERTED_ROLLER_COASTER, vehicle type _4_ACROSS_INVERTED_COASTER_TRAIN
Vehicle entry: ride type MINIATURE_RAILWAY, vehicle type AMERICAN_STYLE_STEAM_TRAIN
Vehicle entry: ride type SUSPENDED_LOOPING_ROLLER_COASTER, vehicle type INVERTED_COASTER_TRAIN
Vehicle entry: ride type SUSPENDED_LOOPING_ROLLER_COASTER, vehicle type FACEOFF_CARS
Vehicle entry: ride type WATER_COASTER, vehicle type WATER_COASTER_BOATS
Vehicle entry: ride type INVERTED_WILD_MOUSE_COASTER, vehicle type SUSPENDED_WILD_MOUSE_CARS
Vehicle entry: ride type AIR_POWERED_VERTICAL_COASTER, vehicle type AIR_POWERED_COASTER_TRAIN
Vehicle entry: ride type JET_SKIS, vehicle type JET_SKIS
Vehicle entry: ride type RAFT_RIDE, vehicle type RAFT_BOATS
Vehicle entry: ride type ENTERPRISE, vehicle type ENTERPRISE_WHEEL
Ride entry: ride type TOILETS
Ride entry: ride type INFORMATION_KIOSK
Ride entry: ride type DRINK_STALL
Scenery entry: item PAGODA_THEMING
Scenery entry: item JURASSIC_THEMING
Scenery entry: item JUNGLE_THEMING
Ride entry: ride type EXOTIC_SEA_FOOD_STALL
Ride entry: ride type ICE_CREAM_STALL
Ride entry: ride type BURGER_BAR
Ride entry: ride type BALLOON_STALL
Ride entry: ride type POPCORN_STALL
Ride entry: ride type HEDGE_MAZE
Ride entry: ride type CIRCUS
Ride entry: ride type MONORAIL
Ride entry: ride type LOG_FLUME
Ride entry: ride type WOODEN_ROLLER_COASTER
Ride entry: ride type STEEL_MINI_ROLLER_COASTER
Ride entry: ride type WOODEN_WILD_MOUSE_ROLLER_COASTER
Scenery entry: item MINE_THEMING
Ride entry: ride type MERRY_GO_ROUND
Ride entry: ride type PIZZA_STALL
Ride entry: ride type MINIATURE_GOLF
Scenery entry: item CLASSICAL_ROMAN_THEMING
Ride entry: ride type CHAIRLIFT
Scenery entry: item EGYPTIAN_THEMING
Scenery entry: item ABSTRACT_THEMING
Ride entry: ride type STEEL_CORKSCREW_ROLLER_COASTER
Ride entry: ride type TWIST
Ride entry: ride type FRIED_CHICKEN_STALL
Ride entry: ride type CYCLE_RAILWAY
Ride entry: ride type LEMONADE_STALL
Ride entry: ride type CANDYFLOSS_STALL
Ride entry: ride type SOUVENIR_STALL
Ride entry: ride type WOODEN_TWISTER_ROLLER_COASTER
Scenery entry: item TOYLAND_THEMING
Ride entry: ride type GHOST_TRAIN
Ride entry: ride type _3D_CINEMA
Scenery entry: item SPACE_THEMING
Ride entry: ride type CROOKED_HOUSE
Ride entry: ride type HOT_DOG_STALL
Ride entry: ride type CHIPS_STALL
Ride entry: ride type BOAT_HIRE
Ride entry: ride type CAR_RIDE
Ride entry: ride type MOTION_SIMULATOR
Ride entry: ride type T_SHIRT_STALL
Ride entry: ride type DOUGHNUT_SHOP
Ride entry: ride type COFFEE_SHOP
Scenery entry: item GRAVEYARD_THEMING
Ride entry: ride type VIRGINIA_REEL
Scenery entry: item MEDIEVAL_THEMING
Ride entry: ride type SPIRAL_SLIDE
Ride entry: ride type BOBSLED_ROLLER_COASTER
Scenery entry: item CREEPY_THEMING
Ride entry: ride type HAUNTED_HOUSE
Ride entry: ride type GO_KARTS
Ride entry: ride type LAUNCHED_FREEFALL
Ride entry: ride type MINE_TRAIN_ROLLER_COASTER
Ride entry: ride type SINGLE_RAIL_ROLLER_COASTER
Ride entry: ride type DODGEMS
Ride entry: ride type MINIATURE_RAILWAY
Ride entry: ride type RAFT_RIDE
Ride entry: ride type OBSERVATION_TOWER
Ride entry: ride type STEEL_ROLLER_COASTER

Uninvented items:
Scenery entry: item JUMPING_FOUNTAINS
Ride entry: ride type AIR_POWERED_VERTICAL_COASTER
Ride entry: ride type ENTERPRISE
Ride entry: ride type TOFFEE_APPLE_STALL
Ride entry: ride type WATER_SLIDE
Ride entry: ride type SUSPENDED_LOOPING_ROLLER_COASTER
Ride entry: ride type RIVER_RIDE
Ride entry: ride type SUSPENDED_ROLLER_COASTER
Ride entry: ride type CYCLE_MONORAIL
Ride entry: ride type SWINGING_SHIP
Ride entry: ride type WOODEN_SIDE_FRICTION_ROLLER_COASTER
Scenery entry: item GARDEN_CLOCK
Ride entry: ride type ROTO_DROP
Ride entry: ride type FERRIS_WHEEL
Ride entry: ride type STAND_UP_STEEL_ROLLER_COASTER
Ride entry: ride type SUSPENDED_MONORAIL
Ride entry: ride type WOODEN_REVERSER_ROLLER_COASTER
Ride entry: ride type RIVER_RAPIDS
Ride entry: ride type REVERSE_FREEFALL_ROLLER_COASTER
Ride entry: ride type SPACE_RINGS
Scenery entry: item MARTIAN_THEMING
Ride entry: ride type INVERTED_WILD_MOUSE_COASTER
Ride entry: ride type JET_SKIS
Ride entry: ride type FLYING_SAUCERS
Ride entry: ride type SUSPENDED_SINGLE_RAIL_ROLLER_COASTER
Ride entry: ride type SWINGING_INVERTER_SHIP
Ride entry: ride type VERTICAL_ROLLER_COASTER
Ride entry: ride type WATER_COASTER
Ride entry: ride type STEEL_WILD_MOUSE_ROLLER_COASTER
Ride entry: ride type HAT_STALL
Ride entry: ride type TOP_SPIN
Ride entry: ride type HEARTLINE_TWISTER_ROLLER_COASTER
Ride entry: ride type FLYING_ROLLER_COASTER
Scenery entry: item SNOW_AND_ICE_THEMING
Ride entry: ride type INVERTED_ROLLER_COASTER
Ride entry: ride type STEEL_TWISTER_ROLLER_COASTER
Scenery entry: item URBAN_THEMING
```

**Review by tupaschoal — COMMENTED** [review]
Thanks for the explanation. Can't we add a test into the unit tests to assure the invention list has the appropriate size?

**Comment by Gymnasiast** [comment]
> Thanks for the explanation. Can't we add a test into the unit tests to assure the invention list has the appropriate size?

There is no way to do that - this is about how much space RCT1 reserves for research items (note that it modifies an array in RCT1.h), not what OpenRCT2 does (which uses a vector, not a fixed-length array). OpenRCT2 research structures have always been big enough to hold all items, from the very beginning.

**Comment by ZeeMaji** [comment]
After testing this branch and comparing it with two tools that can view RCT1 research (RCT-PHP and the RCT Save Game Modifier) i can confirm that LL research seems to be importing correctly now. I've tested 16 of the scenarios from LL which is about half of the ones in the pack, the parks i tested are:

Alton Towers (LL)
Blackpool Pleasure Beach
Heide Park
Iceberg Islands
Volcania
Arid Heights
Razor Rocks
Crater Lake
Vertigo Views
Big Pier 2
Dragon's Cove
Good Knight Park
Wacky Warren
Grand Glacier
Woodworm Park
Micro Park

They all have accurate research now according to RCT-PHP and the RCT Save Game Modifier. Give or take a couple of OpenRCT2 specific things being added or omitted (eg: Jumping fountains/Garden clock scenery "groups" being removed, No ride improvement track piece research, Hyper(twister) coasters as separate ride types.)

**Review by ZehMatt — APPROVED** [review]

**Comment by ZeeMaji** [comment]
It'd probably also make sense to update the PR name/changelog entry to account for the fact that it fixes LL research in general now? Rather than just the incomplete research table import.

File: distribution/changelog.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 0.4.9 (in development)
 ------------------------------------------------------------------------
+- Fix: [#18963] Research table in parks from Loopy Landscapes is imported incompletely.
```
Comment (Gymnasiast) [inline distribution/changelog.txt:3]:
```suggestion
- Fix: [#18963] Research table in parks from Loopy Landscapes is imported incorrectly.
```

**Pull request merged**